When `a` has type `Any`, after `assert a is None`, the rest of the block is considered dead code

Steps to reproduce:

1. Check out [this repo](https://github.com/benkuhn/mypy-repro)
2. Install the packages from `requirements.txt`
3. Run `mypy repro.py` from the repo root.

Expected output:
```
repro.py:4: error: Revealed type is 'Any'
repro.py:6: error: Revealed type is 'builtins.None'
```

Actual output:
```
repro.py:4: error: Revealed type is 'Any'
```

The issue goes away if you disable strict Optional checking.

This is preventing a significant amount of our codebase from being checked right now, so I would appreciate any suggestions for a workaround!

--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -69,16 +69,9 @@ def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool
     TODO: Don't consider callables always overlapping.
     TODO: Don't consider type variables with values always overlapping.
     """
-    # TODO(benkuhn): another option for the fix would be to add the following lines here:
-
-    # # Any overlaps with everything
-    # if isinstance(t, AnyType) or isinstance(s, AnyType):
-    #     return True
-
-    # This passes all the tests locally, and if I understand the mechanics of type overlap
-    # correctly, this would prevent similar classes of bugs from occurring in the future. But I'm
-    # guessing that there's a good reason these lines aren't already there, so I'm defaulting to a
-    # smaller-scoped fix.
+    # Any overlaps with everything
+    if isinstance(t, AnyType) or isinstance(s, AnyType):
+        return True
 
     # Since we are effectively working with the erased types, we only
     # need to handle occurrences of TypeVarType at the top level.
@@ -117,8 +110,6 @@ def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool
         else:
             return False
     if experiments.STRICT_OPTIONAL:
-        if isinstance(t, AnyType) or isinstance(s, AnyType):
-            return True
         if isinstance(t, NoneTyp) != isinstance(s, NoneTyp):
             # NoneTyp does not overlap with other non-Union types under strict Optional checking
             return False

--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -69,17 +69,16 @@ def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool
     TODO: Don't consider callables always overlapping.
     TODO: Don't consider type variables with values always overlapping.
     """
-    # TODO(benkuhn): another option for the fix would be to add the following
-    # lines here:
+    # TODO(benkuhn): another option for the fix would be to add the following lines here:
 
     # # Any overlaps with everything
     # if isinstance(t, AnyType) or isinstance(s, AnyType):
     #     return True
 
-    # This passes all the tests, and if I understand the mechanics of type
-    # overlap correctly, this would prevent similar classes of bugs from
-    # occurring in the future. But I'm guessing that there's a good reason these
-    # lines aren't already there, so I'm defaulting to a smaller-scoped fix.
+    # This passes all the tests locally, and if I understand the mechanics of type overlap
+    # correctly, this would prevent similar classes of bugs from occurring in the future. But I'm
+    # guessing that there's a good reason these lines aren't already there, so I'm defaulting to a
+    # smaller-scoped fix.
 
     # Since we are effectively working with the erased types, we only
     # need to handle occurrences of TypeVarType at the top level.

--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -69,6 +69,9 @@ def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool
     TODO: Don't consider callables always overlapping.
     TODO: Don't consider type variables with values always overlapping.
     """
+    # If either side is Any, there is definitely overlap.
+    if isinstance(t, AnyType) or isinstance(s, AnyType):
+        return True
     # Since we are effectively working with the erased types, we only
     # need to handle occurrences of TypeVarType at the top level.
     if isinstance(t, TypeVarType):

--- a/mypy/meet.py
+++ b/mypy/meet.py
@@ -69,9 +69,18 @@ def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool
     TODO: Don't consider callables always overlapping.
     TODO: Don't consider type variables with values always overlapping.
     """
-    # If either side is Any, there is definitely overlap.
-    if isinstance(t, AnyType) or isinstance(s, AnyType):
-        return True
+    # TODO(benkuhn): another option for the fix would be to add the following
+    # lines here:
+
+    # # Any overlaps with everything
+    # if isinstance(t, AnyType) or isinstance(s, AnyType):
+    #     return True
+
+    # This passes all the tests, and if I understand the mechanics of type
+    # overlap correctly, this would prevent similar classes of bugs from
+    # occurring in the future. But I'm guessing that there's a good reason these
+    # lines aren't already there, so I'm defaulting to a smaller-scoped fix.
+
     # Since we are effectively working with the erased types, we only
     # need to handle occurrences of TypeVarType at the top level.
     if isinstance(t, TypeVarType):
@@ -109,6 +118,8 @@ def is_overlapping_types(t: Type, s: Type, use_promotions: bool = False) -> bool
         else:
             return False
     if experiments.STRICT_OPTIONAL:
+        if isinstance(t, AnyType) or isinstance(s, AnyType):
+            return True
         if isinstance(t, NoneTyp) != isinstance(s, NoneTyp):
             # NoneTyp does not overlap with other non-Union types under strict Optional checking
             return False

--- a/test-data/unit/check-optional.test
+++ b/test-data/unit/check-optional.test
@@ -77,6 +77,15 @@ else:
   reveal_type(x)  # E: Revealed type is 'builtins.int'
 [builtins fixtures/bool.pyi]
 
+[case testAnyCanBeNone]
+from typing import Optional, Any
+x = None  # type:  Any
+if x is None:
+  reveal_type(x)  # E: Revealed type is 'builtins.None'
+else:
+  reveal_type(x)  # E: Revealed type is 'Any'
+[builtins fixtures/bool.pyi]
+
 [case testOrCases]
 from typing import Optional
 x = None  # type: Optional[str]

